SmartDeviceLink/SDLSecondaryTransportManager.m

```diff
@@ -0,0 +1,806 @@
+    if (startServiceACK.header.serviceType != SDLServiceTypeRPC) {
+        return;
+    }
+
+    SDLLogV(@"Received Start Service ACK header of RPC service on primary transport");
+
+    // keep header to acquire Session ID
+    self.primaryRPCHeader = startServiceACK.header;
+
+    SDLControlFramePayloadRPCStartServiceAck *payload = [[SDLControlFramePayloadRPCStartServiceAck alloc] initWithData:startServiceACK.payload];
+
+    NSDictionary<NSString *, id> *userInfo = @{SDLNotificationUserInfoObject: payload};
+    [[NSNotificationCenter defaultCenter] postNotificationName:SDLStartSecondaryTransportManagerNotification object:self userInfo:userInfo];
+}
+
+// called from protocol's _reeiveQueue of Primary Transport
+- (void)handleTransportEventUpdateMessage:(SDLProtocolMessage *)transportEventUpdate {
+
+    SDLControlFramePayloadTransportEventUpdate *payload = [[SDLControlFramePayloadTransportEventUpdate alloc] initWithData:transportEventUpdate.payload];
+    SDLLogV(@"Transport Config Update: %@", payload);
+
+    NSDictionary<NSString *, id> *userInfo = @{SDLNotificationUserInfoObject: payload};
+    [[NSNotificationCenter defaultCenter] postNotificationName:SDLTransportEventUpdateNotification object:self userInfo:userInfo];
+}
+
+@end
+
+@interface SDLSecondaryTransportManager ()
+
+/// state of this manager
+@property (strong, nonatomic, readwrite) SDLStateMachine *stateMachine;
+/// dedicated queue that the state machine will run on
+@property (copy, nonatomic) dispatch_queue_t stateMachineQueue;
+
+/// instance of the protocol that runs on primary transport
+@property (weak, nonatomic) SDLProtocol *primaryProtocol;
+/// a sub-class to catch Start Service ACK and Transport Config Update frames
+@property (strong, nonatomic) PrimaryProtocolListener *primaryProtocolListener;
+
+/// Selected type of secondary transport. If 'SDLTransportSelectionDisabled' then secondary transport is disabled.
+@property (assign, nonatomic) SDLTransportSelection transportSelection;
+/// instance of the transport for secondary transport
+@property (nullable, strong, nonatomic) id<SDLTransportType> secondaryTransport;
+/// instance of the protocol that runs on secondary transport
+@property (nullable, strong, nonatomic) SDLProtocol *secondaryProtocol;
+/// timer to check Register Secondary Transport ACK response on secondary transport
+@property (strong, nonatomic, nullable) SDLTimer *registerTransportTimer;
+
+// a listener to notify protocol change
+@property (weak, nonatomic) id<SDLStreamingProtocolListener> streamingProtocolListener;
+
+/// Configuration sent by system; list of transports that are allowed to carry audio service
+@property (strong, nonatomic, nullable) NSArray<SDLTransportClassBox *> *transportsForAudioService;
+/// Configuration sent by system; list of transports that are allowed to carry video service
+@property (strong, nonatomic, nullable) NSArray<SDLTransportClassBox *> *transportsForVideoService;
```

Comment:
Hmm I prefer current approach. When Core doesn't send the list, the array is null. When it sends an empty list, the array is non-null and is empty.
I could make the array to be always non-null but I think that would cause more confusion.

Reply:
It may closer match to what core sends, but I think it complicates the logic. When we ask the question "what are the transports for audio" we don't expect `NULL`, we expect the answer to be "none". Or, perhaps more clearly, if we ask, "how many transports for audio do we have", we don't expect the answer to be `NULL`, it would be `0` if there are none. Not `NULL` in some cases and `0` in others.

Reply:
OK, understood. I updated the code in commit b32fac3ea910101e1856f927bd573bdf0fdfe4f2.